### PR TITLE
fix(Core/Spells): Fixed implementation of CAST_FLAG_POWER_LEFT_SELF f…

### LINKED ISSUE #10820: Judgement of Light, Judgement of Wisdom healing/mana regen proc
### What client do you play on?

enUS

### Faction

- [X] Alliance
- [X] Horde

### Content Phase:

- [X] Generic
- [ ] 1-19
- [ ] 20-29
- [ ] 30-39
- [ ] 40-49
- [ ] 50-59
- [ ] 60

Originally reported:  https://github.com/chromiecraft/chromiecraft/issues/3100

### Current Behaviour

When you attack a target affected by Judgement of Light or Judgement of Wisdom your mana stops regenerating. 

Proc might be considered a cast so if you don't have mana regen while casting you don't get any while the healing (or proc from Judgement of Wisdom) happens.

Judgement of Light:

https://user-images.githubusercontent.com/91145640/154801597-4f442d3a-57d7-4eec-9d99-823e95401c44.mp4

Judgement of Wisdom:

https://user-images.githubusercontent.com/91145640/154801776-e3c2d36e-e527-4d85-bbf2-72f18a8be634.mp4

### Expected Blizzlike Behaviour

Mana regeneration shouldn't stop.

### Source

_No response_

### Steps to reproduce the problem

1. Use Judgement of Light/Wisdom without any MP5.
2. Check mana regen on judgment proc.

### Extra Notes

Bug found by Omniknight.

Issue regarding the current JoL, JoW debuff behaviour.

Describing Issue: Once party/raid targets auto attack the target with JoL/JoW debuff, when the proc effects for healing/mana regen activates, the proc is registered as an instant cast/spell, mitigating mana regeneration by around 45%, as that is the issue once spells or casts have been used. Therefore, when Judgment debuff healing/mana regen procs on auto attack, players who auto attacked the debuffed target will stop regenerating mana completely if they have no mana regen and players with some or high mana regen will experience a loss of mana regeneration of around 45% as is stated for spells and casts.

Possible issue cause: Cause could be that the proc is being registered as the instant cast/spell, therefore mitigating mana regen on auto attack proc for all players involved in the raid/party.
### AC rev. hash/commit

https://github.com/chromiecraft/azerothcore-wotlk/commit/f82c60526b9955c697b1ee804a8e7613f62c1568

### Operating system

Windows 10.

### Modules

- [mod-ah-bot](https://github.com/azerothcore/mod-ah-bot)
- [mod-cfbg](https://github.com/azerothcore/mod-cfbg)
- [mod-chat-transmitter](https://github.com/azerothcore/mod-chat-transmitter)
- [mod-chromie-xp](https://github.com/azerothcore/mod-chromie-xp)
- [mod-desertion-warnings](https://github.com/azerothcore/mod-desertion-warnings)
- [mod-duel-reset](https://github.com/azerothcore/mod-duel-reset)
- [mod-eluna-lua-engine](https://github.com/azerothcore/mod-eluna-lua-engine)
- [mod-ip-tracker](https://github.com/azerothcore/mod-ip-tracker)
- [mod-low-level-arena](https://github.com/azerothcore/mod-low-level-arena)
- [mod-multi-client-check](https://github.com/azerothcore/mod-multi-client-check)
- [mod-pvp-titles](https://github.com/azerothcore/mod-pvp-titles)
- [mod-pvpstats-announcer](https://github.com/azerothcore/mod-pvpstats-announcer)
- [mod-queue-list-cache](https://github.com/azerothcore/mod-queue-list-cache)
- [mod-server-auto-shutdown](https://github.com/azerothcore/mod-server-auto-shutdown)
- [mod-save-inventory](https://github.com/Nyeriah/mod-save-inventory)
- [lua-carbon-copy](https://github.com/55Honey/Acore_CarbonCopy)
- [lua-exchange-npc](https://github.com/55Honey/Acore_ExchangeNpc)
- [lua-custom-worldboss](https://github.com/55Honey/Acore_CustomWorldboss)
- [lua-level-up-reward](https://github.com/55Honey/Acore_LevelUpReward)
- [lua-recruit-a-friend](https://github.com/55Honey/Acore_RecruitAFriend)
- [lua-send-and-bind](https://github.com/55Honey/Acore_SendAndBind)
- [lua-temp-announcements](https://github.com/55Honey/Acore_TempAnnouncements)
- [lua-zonecheck](https://github.com/55Honey/acore_Zonecheck)
- [lua-zone-debuff](https://github.com/55Honey/Acore_ZoneDebuff)

### Customizations

None

### Server

ChromieCraft

### PR DESCRIPTION
**Pull request by UltraNix**
…lag.

Do not send power update packets when gaining power via energize spell effects to match retail behavior (source: TrinityCore).
Fixes #10820

## Issues Addressed:

- Closes #10820
- Closes https://github.com/chromiecraft/chromiecraft/issues/3100

## Tests Performed:

- Tested ingame.

## How to Test the Changes:

Use Judgement of Light/Wisdom without any MP5.
Check mana regen on judgment proc.

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

**Comment by mpfans** [comment]
@Dejanos test it,please

**Comment by Dejanos** [comment]
Just tested it on live server. My shaman still loses all spirit mana regen when healing Judgement procs on target dummy on his auto attacks, when the effect stops being active after 2-3 seconds, he continues regaining mana. (Base spirit mana regeneration currently should be about the same for all classes, around 116 spirit, or 49-48 mana per 5 seconds).

**Comment by UltraNix** [comment]
@Dejanos Please, record a videos about this. I see no bug currently with this PR.

**Comment by Dejanos** [comment]
[[](url)](https://streamable.com/wp6hdk)

Upon further testing I do think it is working as intended now, just mana regen gets a little big choppy visual representation in combat, but I could almost swear that was not the case the last time I tested it. In the beginning of the video I hover over Spirit tab, and then Mana Regen, to show I currently have 0 mana regen and only base spirit as regeneration, then I test it with Lesser Healing Wave and see that mana regen actually stops as it should for a few seconds, because all I have is spirit. 
After that I went on to auto attack on a dummy and only experienced visual choppy regeneration, but it continued with no problem for me, so I am sorry if this caused any further unnecessary delays from my part. 

As it seems from the video and my current testing, the regeneration indeed continues as it should upon Judgement auto attack healing effect. Keep up the good job and thank you for the fix.

**Review by Nefertumm — APPROVED** [review]

File: src/server/game/Spells/Spell.cpp [hunk 1]
```diff
@@ -4461,10 +4461,12 @@ void Spell::SendSpellStart()
 
     if (m_spellInfo->HasAttribute(SPELL_ATTR0_USES_RANGED_SLOT) || m_spellInfo->HasAttribute(SPELL_ATTR0_CU_NEEDS_AMMO_DATA))
         castFlags |= CAST_FLAG_PROJECTILE;
-    if ((m_caster->GetTypeId() == TYPEID_PLAYER ||
-            (m_caster->GetTypeId() == TYPEID_UNIT && m_caster->IsPet()))
-            && m_spellInfo->PowerType != POWER_HEALTH)
+
+    if ((m_caster->GetTypeId() == TYPEID_PLAYER || (m_caster->GetTypeId() == TYPEID_UNIT && m_caster->IsPet()))
+        && m_spellInfo->PowerType != POWER_HEALTH && m_powerCost != 0)
```
Comment (r0m1ntik) [inline src/server/game/Spells/Spell.cpp:4466]: ```m_powerCost != 0```
[bug](https://github.com/azerothcore/azerothcore-wotlk/issues/11233)

File: src/server/game/Spells/Spell.cpp [hunk 2]
```diff
@@ -4518,10 +4520,11 @@ void Spell::SendSpellGo()
     if (m_spellInfo->HasAttribute(SPELL_ATTR0_USES_RANGED_SLOT) || m_spellInfo->HasAttribute(SPELL_ATTR0_CU_NEEDS_AMMO_DATA))
         castFlags |= CAST_FLAG_PROJECTILE;                        // arrows/bullets visual
 
-    if ((m_caster->GetTypeId() == TYPEID_PLAYER ||
-            (m_caster->GetTypeId() == TYPEID_UNIT && m_caster->IsPet()))
-            && m_spellInfo->PowerType != POWER_HEALTH)
-        castFlags |= CAST_FLAG_POWER_LEFT_SELF; // should only be sent to self, but the current messaging doesn't make that possible
+    if ((m_caster->GetTypeId() == TYPEID_PLAYER || (m_caster->GetTypeId() == TYPEID_UNIT && m_caster->IsPet()))
+        && m_spellInfo->PowerType != POWER_HEALTH && m_powerCost != 0) // should only be sent to self, but the current messaging doesn't make that possible
```
Comment (r0m1ntik) [inline src/server/game/Spells/Spell.cpp:4524]: ```m_powerCost != 0```
[bug](https://github.com/azerothcore/azerothcore-wotlk/issues/11233)

**Review by r0m1ntik — COMMENTED** [review]
[https://github.com/azerothcore/azerothcore-wotlk/issues/11233](https://github.com/azerothcore/azerothcore-wotlk/issues/11233)

**Pull request merged**